### ISSUE TITLE
`restoreRecentDocuments()` causes IINA to hang when network storage is unavailable

### ISSUE THREAD

**Issue by LeoNatan** [iina/iina#5299]
### IINA Version

1.3.5

### Release Channel

Release

### macOS Version

15.2

### Steps to Reproduce

- Launch IINA while network storage is unavailable

### Expected Behavior

App should launch

### Actual Behavior

App hangs on launch, bouncing in the dock. Running spindump, the app is stuck synchronously attempting to mount a network share which isn't available:

```
  Thread 0x5d6c    DispatchQueue "com.apple.main-thread"(1)    1001 samples (1-1001)    priority 46 (base 46)
  1001  start + 2840 (dyld + 25204) [0x19361c274]
    1001  main + 12 (IINA + 1014528) [0x102f67b00]
      1001  NSApplicationMain + 888 (AppKit + 18516) [0x1975b6854]
        1001  -[NSApplication run] + 252 (AppKit + 188284) [0x1975dff7c]
          1001  -[NSApplication finishLaunching] + 248 (AppKit + 189104) [0x1975e02b0]
            1001  -[NSNotificationCenter postNotificationName:object:userInfo:] + 88 (Foundation + 40916) [0x194c01fd4]
              1001  _CFXNotificationPost + 732 (CoreFoundation + 266140) [0x193a47f9c]
                1001  _CFXRegistrationPost + 436 (CoreFoundation + 1061204) [0x193b0a154]
                  1001  ___CFXRegistrationPost_block_invoke + 88 (CoreFoundation + 1061388) [0x193b0a20c]
                    1001  __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148 (CoreFoundation + 467808) [0x193a79360]
                      1001  @objc AppDelegate.applicationWillFinishLaunching(_:) + 100 (IINA + 1005028) [0x102f655e4]
                        1001  AppDelegate.applicationWillFinishLaunching(_:) + 1640 (IINA + 992040) [0x102f62328]
                          1001  specialized AppDelegate.restoreRecentDocuments() + 724 (IINA + 1023816) [0x102f69f48]
                            1001  URL.init(resolvingBookmarkData:options:relativeTo:bookmarkDataIsStale:) + 168 (Foundation + 8002844) [0x195399d1c]
                              1001  specialized @nonobjc NSURL.init(resolvingBookmarkData:options:relativeTo:bookmarkDataIsStale:) + 140 (Foundation + 8122624) [0x1953b7100]
                                1001  -[NSURL(NSURL) initByResolvingBookmarkData:options:relativeToURL:bookmarkDataIsStale:error:] + 420 (Foundation + 621196) [0x194c8fa8c]
                                  1001  _CFURLCreateByResolvingBookmarkData + 148 (CoreServicesInternal + 68568) [0x1973f1bd8]
                                    1001  createByResolvingBookmarkDataInternal(__CFAllocator const*, __CFData const*, unsigned long, __CFURL const*, __CFArray const*, unsigned char*, __CFError**, __CFURL const**) + 1276 (CoreServicesInternal + 70416) [0x1973f2310]
                                      1001  CFURLCreateByResolvingDataInBookmark(__CFAllocator const*, BookmarkData&, unsigned int, unsigned long, __CFArray const*, unsigned char*, __CFError**, __CFURL const**) + 3868 (CoreServicesInternal + 75872) [0x1973f3860]
                                        1001  NetFSMountURLSync + 252 (NetFS + 38880) [0x1b8d247e0]
                                          1001  NAConnectToServerSync + 116 (NetAuth + 12432) [0x1a53e9090]
                                            1001  _NetAuthConnectToServerSync + 304 (NetAuth + 12780) [0x1a53e91ec]
                                              1001  NAAA_MountURL + 236 (NetAuth + 43708) [0x1a53f0abc]
                                                1001  mach_msg + 24 (libsystem_kernel.dylib + 4732) [0x19395b27c]
                                                  1001  mach_msg_overwrite + 480 (libsystem_kernel.dylib + 39640) [0x193963ad8]
                                                    1001  mach_msg2_trap + 8 (libsystem_kernel.dylib + 3892) [0x19395af34]
                                                     *1001  ??? (kernel.release.t6031 + 240616) [0xfffffe00086eabe8]
```

Not sure what "workaround" `restoreRecentDocuments()` is necessary for, but doing that work synchronously on the main thread is never a good idea. I guess nobody thought IINA is used with network shares.

### Crash Report or Process Sample

[Spindump.txt](https://github.com/user-attachments/files/18045290/Spindump.txt)

### How often can you reproduce the issue

Always

### Required Actions

- [X] I tried my best to search from the existing issues (including closed issues) and didn't find duplicated issues.

### Optional Actions

- [ ] I tested with the Nightly build and confirmed the issue is not fixed.
- [ ] I tested with the latest version of mpv and confirmed mpv doesn't have this issue.

**Comment by low-batt** [comment]
This is issue #4973. The workaround is given in [this comment](https://github.com/iina/iina/issues/4973#issuecomment-2181352809).

Indeed as you identified, this is due to an attempt to workaround issue #4688. Unfortunately the side effect of macOS waiting on a very long timeout if the list of recents contains a reference to a file on a device that is unreachable was not encountered during testing (NAS was always online). As we could not find a way to tell macOS not to do that, that workaround has been disabled in the development version of IINA.

The lost of the recent documents started happening with macOS Sonoma. Later on Apple added that change to macOS Ventura in an update. In IINA this means all the previously played videos will disappear from the welcome window when macOS clears the list.

Experiments show the macOS `sharedfilelistd` daemon is clearing the list if the code signature of the executable changes. Users who only use the release and beta versions of IINA should not encounter the problem. The workaround was added over worries about users who test IINA nightly builds having this history erased. 

Hard to find information on the net about this. Too many hits about past problems that were specific to apps miss-managing the recents list. I haven't searched recently, but back then I only found one post where someone was encountering this. Maybe this macOS change was some sort of attempt at security? If so, deleting user data seems like a bad choice and not useful from a security standpoint since the property file containing the list is not protected and can be directly read by programs with access to the user's directory tree. Reports from users seem to indicate the list is sometimes being cleared for no apparent reason. So possibly this macOS change is also unreliable.

As this problem is very irritating for developers for whom the list is constantly being cleared, the workaround is still in the code base, but disabled by default.

**Comment by LeoNatan** [comment]
Sounds like a privacy feature, where a software impersonating another would be unable to get a list of files. The workaround makes sense, and it has been since fixed (disabled by default), but no public release with that fix.

**Comment by LeoNatan** [comment]
BTW, why not just sign and notarize nightly builds?

**Comment by low-batt** [comment]
We discussed signing the nightly builds, but rejected that idea over security concerns about an automated build having access to the signing certificates.

We also considered using a different bundle ID for the nightly build. But that causes problems for testing scenarios as the nightly will have its own configuration settings. This idea was rejected as too disruptive.

And yes we are late publishing a release. I'm hoping we will be releasing a beta soon.

It does seem like Apple made this change to improve privacy or address a security concern. But it doesn't accomplish that because the file containing the list is:
```
~/Library/Application Support/com.apple.sharedfilelist/com.apple.LSSharedFileList.ApplicationRecentDocuments/com.colliderli.iina.sfl3
```
Which is accessible to any app that has read access to the user's files. A malicious app just has to directly access that file instead of using the AppKit API. I'm definitely confused about why Apple made this change.

**Comment by LeoNatan** [comment]
There are approaches to having the private key private in the CI system, and only signing builds from master (and not e.g. PR builds). Without signing and having the exact entitlement chain, macOS behaves oddly, as you see in this case.

**Comment by low-batt** [comment]
I can't speak with authority to the concerns about adding signing to the CI system. I'm a junior IINA developer and not involved in that part of the project. From what I recall the nightly builds were not using GitHub at that time. So the security concern may have had to do with the old build system.

Technically the nightly builds are not unsigned. They are signed, but with an [ad-hoc](https://developer.apple.com/documentation/security/seccodesignatureflags/adhoc) signature.

<details>
<summary><b>Code Signature:</b></summary>

```text
low-batt@gag nightly$ codesign --display --verbose IINA.app
Executable=/Users/low-batt/Downloads/sw/io/iina/nightly/IINA.app/Contents/MacOS/IINA
Identifier=com.colliderli.iina
Format=app bundle with Mach-O universal (x86_64 arm64)
CodeDirectory v=20500 size=55604 flags=0x10002(adhoc,runtime) hashes=1727+7 location=embedded
Signature=adhoc
Info.plist entries=42
TeamIdentifier=not set
Runtime Version=15.1.0
Sealed Resources version=2 rules=13 files=1227
Internal requirements count=0 size=12
```
</details>

Good point about entitlements. The entitlements used by the official release versions of IINA are:
- [com.apple.security.cs.allow-unsigned-executable-memory](https://developer.apple.com/documentation/bundleresources/entitlements/com.apple.security.cs.allow-unsigned-executable-memory)
- [com.apple.security.cs.disable-library-validation](https://developer.apple.com/documentation/bundleresources/entitlements/com.apple.security.cs.disable-library-validation)

Use of `com.apple.security.cs.allow-unsigned-executable-memory` was added as a workaround for issue  #3551. We are unable to use `com.apple.security.cs.allow-jit` due to [luajit2 #145](https://github.com/openresty/luajit2/issues/145).

As nightly builds are debuggable they also have the  `com.apple.security.get-task-allow` entitlement.

From a quick search I was not able to find a list of restricted entitlements that are not available to ad-hoc signed apps. I'm guessing those are some of the more risky ones like access to Apple Pay.

**Comment by LeoNatan** [comment]
I ended up downloading a nightly build, and on Sequoia, running that is a terrible experience. There isn't much difference between an adhoc-signed and unsigned binary when the bundle is quarantined. Apple has made these operating systems so obtuse, Vista is a joy to use in comparison. pRiVaCy AnD sEcUriTy

**Comment by low-batt** [comment]
I too was nailed by Apple changing the procedure to allow apps to run in macOS Sequoia. Due to that I started working on a FAQ to explain what a nightly build is and how to run it. I've not had a chance to finish it yet.

The nightly build contains changes from 120+ completed tasks. Only a couple of tasks remain to be completed before the first beta is released. The nightly contains significant changes to address thread coordination problems that had been judged risky and postponed to a release that uses betas. There is also major upgrades to the versions of the libraries IINA uses. Lots of other changes.

There have been issues with Sequoia. I forget the problem right now, but I was unable to reproduce it and at some point the user reporting it rebooted macOS and it no longer reproduced for them. A user just reported a problem with the `Use legacy full screen` setting. I've reproduced it and confirmed things work under Ventura. As that was just reported I've not had a chance to investigate yet and see what Apple has broken. I do not have a good feeling about Sequoia.

Sorry again about the hang in 1.3.5.

**Comment by LeoNatan** [comment]
No worries, poop happens. Thank you for maintaining the project!

**Comment by LeoNatan** [comment]
`xattr -d com.apple.quarantine` is enough to run the nightly. I’ve been meaning to create a small tool that looks for stuff in Downloads and Applications and removes quarantine automatically.

**Comment by jadenquinn** [comment]
The app shouldn't be trying to mount the damn share at all. It should ask you if you want to before trying to. Duh.

**Comment by low-batt** [comment]
The app doesn't explicitly ask for a share to be mounted. IINA is performing operations on bookmarks not expecting macOS to ask for the share to be mounted.

This is issue #4973. The workaround is given in [this comment](https://github.com/iina/iina/issues/4973#issuecomment-2181352809).

**Comment by LeoNatan** [comment]
We are at the end of January and there is still no release with this fix.

**Comment by low-batt** [comment]
IINA has been in final release testing for the first beta release of the upcoming major release since December. More work has been put into testing this time and as a result multiple serious problems were detected. Some required coordination with another project to obtain a fix. A second internal release candidate was just rejected after additional problems were found. We are currently investigating and fixing those problems. We are hoping that once these problems are fixed a third release candidate will pass internal testing and we can release the first beta.

**Comment by low-batt** [comment]
This was fixed in IINA 1.4.0-beta1.

**Comment by jadenquinn** [comment]
Awesome! Updated.
